scripts/fetch-products.js

```diff
@@ -0,0 +1,118 @@
+require('dotenv').config();
+
+const fs = require('fs');
+const path = require('path');
+const ethers = require('ethers');
+const fetch = require('node-fetch');
+
+const coverAbi = require('../src/contracts/abis/Cover.json');
+const addresses = require('../src/contracts/addresses.json');
+const logos = require('../src/products/logos.json');
+
+const { PROVIDER_URL, IPFS_GATEWAY_URL } = process.env;
+
+const ipfsURL = ipfsHash => `${IPFS_GATEWAY_URL}/ipfs/${ipfsHash}`;
+
+const fetchProductTypes = async cover => {
+  const eventFilter = cover.filters.ProductTypeSet();
+  const events = await cover.queryFilter(eventFilter);
+
+  // using sort and reduce to deduplicate events and get the latest ipfs hash
+  const ipfsHashes = events
+    .sort((a, b) => a.logIndex - b.logIndex)
+    .reduce((acc, event) => {
+      const id = event.args.id.toNumber();
+      const coverWordingURL = ipfsURL(event.args.ipfsMetadata);
+      return { ...acc, [id]: { id, coverWordingURL } };
+    }, {});
+
+  const productTypes = [];
+
+  for (const item of Object.values(ipfsHashes)) {
+    const name = await cover.productTypeNames(item.id);
+    const { gracePeriod, claimMethod } = await cover.productTypes(item.id);
+    productTypes.push({ ...item, name, gracePeriod, claimMethod });
+  }
+
+  console.log('productTypes', productTypes);
+
+  return productTypes;
+};
+
+const fetchProducts = async cover => {
+  const eventFilter = cover.filters.ProductSet();
+  const events = await cover.queryFilter(eventFilter);
+
+  const ipfsHashes = events
+    .sort((a, b) => a.logIndex - b.logIndex)
+    .reduce((acc, event) => {
+      const id = event.args.id.toNumber();
+      const ipfsHash = event.args.ipfsMetadata;
+      return { ...acc, [id]: { id, ipfsHash } };
+    }, {});
+
+  const productsCount = await cover.productsCount();
+  const ids = Array.from(Array(productsCount.toNumber()).keys());
+  const batches = [];
+
+  while (ids.length) {
+    batches.push(ids.splice(0, 50));
+  }
+
+  const products = [];
+
+  for (const batch of batches) {
+    const promises = batch.map(async id => {
+      const { productType, isDeprecated, useFixedPrice } = await cover.products(id);
+      const name = await cover.productNames(id);
+      const { ipfsHash } = ipfsHashes[id];
+      const metadata = ipfsHash === '' ? {} : await fetch(ipfsURL(ipfsHash)).then(res => res.json());
+
+      if (logos[id] === undefined) {
```

Comment:
I think the `logos[id] === undefined` and the `ipfsHash === ''` checks could be changed into falsy checks like `!logos[id]` and `!ipfsHash` to catch some empty string cases here and there (especially for env vars) and make it a tiny bit more readable. If you are doing these checks so specific for a reason than by all means leave them in